File: packages/node/src/aad.ts
import { TextEncoder } from "node:util";
import { 
  NsForbiddenError, 
  NsCollisionError,
  X402HeaderError,
  X402PayloadError,
  X402ExtensionUnapprovedError,
  X402ExtensionDuplicateError,
  X402ExtensionPayloadError
} from "./errors.js";
import { X402Extension, isApprovedExtensionHeader, canonicalizeExtensionHeader } from "./extensions.js";

export type X402Core = {
  header: string; // "X-Payment" | "X-Payment-Response" (case-insensitive input)
  payload: Record<string, any>;
  // Additional KV allowed and included in canonicalization
  [k: string]: any;
};

const enc = new TextEncoder();

function deepCanonicalize(value: any): any {
  if (value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) return value.map(deepCanonicalize);
  const keys = Object.keys(value).sort();
  const out: any = {};
  for (const k of keys) out[k] = deepCanonicalize(value[k]);
  return out;
}

function canonicalizeHeaderCase(h: string): "X-Payment" | "X-Payment-Response" | "" {
  const s = String(h).toLowerCase();
  if (s === "x-payment") return "X-Payment";
  if (s === "x-payment-response") return "X-Payment-Response";
  if (s === "") return "";
  throw new X402HeaderError("X402_HEADER");
}

export function validateX402Core(x: any): X402Core {
  const header = canonicalizeHeaderCase(x?.header);
  const payload = x?.payload;
  if (header !== "" && (!payload || typeof payload !== "object" || Array.isArray(payload) || Object.keys(payload).length === 0)) {
    throw new X402PayloadError("X402_PAYLOAD");
  }
  const extra: any = {};
  for (const k of Object.keys(x || {})) {
    if (k === "header" || k === "payload") continue;
    extra[k] = x[k];
  }
  return { header, payload, ...extra };
}

function canonicalJson(obj: Record<string, any>): string {
  return JSON.stringify(deepCanonicalize(obj));
}

export function buildCanonicalAad(
  namespace: string,
  payload: {
    request?: Record<string, any>;
    response?: Record<string, any>;
    x402?: X402Core;
  },
  extensions?: X402Extension[]
): {
  aadBytes: Uint8Array;
  x402Normalized?: X402Core;
  requestNormalized?: Record<string, any>;
  responseNormalized?: Record<string, any>;
  extensionsNormalized?: X402Extension[];
} {
  if (!namespace || namespace.toLowerCase() === "x402") throw new NsForbiddenError("NS_FORBIDDEN");
  
  const { request, response, x402 } = payload;
  let primaryJson = "";
  let x402Normalized, requestNormalized, responseNormalized;

  if (x402) {
    const x = validateX402Core(x402);
    primaryJson = canonicalJson(x);
    x402Normalized = JSON.parse(primaryJson);
  } else if (request) {
    primaryJson = canonicalJson(request);
    requestNormalized = JSON.parse(primaryJson);
  } else if (response) {
    primaryJson = canonicalJson(response);
    responseNormalized = JSON.parse(primaryJson);
  }

  let extensionsNormalized;
  let extensionsJson = "";
  if (Array.isArray(extensions) && extensions.length > 0) {
    const seen = new Set<string>();
    const exts: X402Extension[] = [];
    for (const e of extensions as any[]) {
      const hdr = String(e?.header || "");
      if (!isApprovedExtensionHeader(hdr)) throw new X402ExtensionUnapprovedError("X402_EXTENSION_UNAPPROVED");
      const canonHdr = canonicalizeExtensionHeader(hdr);
      if (seen.has(canonHdr.toLowerCase())) throw new X402ExtensionDuplicateError("X402_EXTENSION_DUPLICATE");
      const extPayload = e?.payload;
      if (!extPayload || typeof extPayload !== "object" || Array.isArray(extPayload) || Object.keys(extPayload).length === 0) {
        throw new X402ExtensionPayloadError("X402_EXTENSION_PAYLOAD");
      }
      const extExtra: any = {};
      for (const k2 of Object.keys(e)) if (k2 !== "header" && k2 !== "payload") extExtra[k2] = e[k2];
      exts.push({ header: canonHdr, payload: extPayload, ...extExtra });
      seen.add(canonHdr.toLowerCase());
    }
    // Sort extensions by header (case-insensitive)
    exts.sort((a, b) => a.header.toLowerCase().localeCompare(b.header.toLowerCase()));
    extensionsNormalized = exts.map((e) => deepCanonicalize(e));
    extensionsJson = canonicalJson(extensionsNormalized);
  }
  const prefix = `${namespace}|v1|`;
  const suffix = extensionsJson ? `|${extensionsJson}` : "|";
  const full = prefix + primaryJson + suffix;
  return { 
    aadBytes: enc.encode(full), 
    x402Normalized, 
    requestNormalized, 
    responseNormalized, 
    extensionsNormalized 
  };
}

export function canonicalAad(
  namespace: string, 
  payload: {
    request?: Record<string, any>;
    response?: Record<string, any>;
    x402?: X402Core;
  },
  extensions?: X402Extension[]
): Uint8Array {
  return buildCanonicalAad(namespace, payload, extensions).aadBytes;
}

// V2 refactor types
export type PrivateHeaderEntry = {
  header: string;
  value: any;
  [k: string]: any;
};

export function canonicalizeCoreHeaderName(h: string): "X-Payment" | "X-Payment-Response" | "" | string {
  const s = String(h || "");
  if (s === "") return "";
  const sl = s.toLowerCase();
  if (sl === "x-payment") return "X-Payment";
  if (sl === "x-payment-response") return "X-Payment-Response";
  return s;
}

function canonicalJsonCompact(obj: any): string {
  return JSON.stringify(deepCanonicalize(obj));
}

export function buildCanonicalAadHeadersBody(
  namespace: string,
  privateHeaders?: PrivateHeaderEntry[] | undefined,
  privateBody?: Record<string, any> | undefined
): {
  aadBytes: Uint8Array;
  headersNormalized: Array<{ header: string; value: any; [k: string]: any }>;
  bodyNormalized: Record<string, any>;
} {
  if (!namespace || namespace.toLowerCase() === "x402") throw new NsForbiddenError("NS_FORBIDDEN");

  // Normalize headers
  const headersIn = Array.isArray(privateHeaders) ? privateHeaders : [];
  const seen = new Set<string>();
  const headersNormalized = headersIn.map((e) => {
    const rawHdr = String(e?.header ?? "");
    let hdr = canonicalizeCoreHeaderName(rawHdr);
    if (hdr !== "X-Payment" && hdr !== "X-Payment-Response" && hdr !== "") {
      // extension header path
      if (!isApprovedExtensionHeader(hdr)) {
        throw new X402ExtensionUnapprovedError("X402_EXTENSION_UNAPPROVED");
      }
      hdr = canonicalizeExtensionHeader(hdr);
    }
    const key = hdr.toLowerCase();
    if (seen.has(key)) throw new X402ExtensionDuplicateError("X402_EXTENSION_DUPLICATE");
    seen.add(key);
    const out: any = { ...e };
    out.header = hdr;
    // Canonicalize value structure for AAD stability
    out.value = deepCanonicalize(e?.value);
    return out;
  });
  headersNormalized.sort((a, b) => a.header.toLowerCase().localeCompare(b.header.toLowerCase()));

  // Normalize body
  const bodyNormalized = privateBody ? (deepCanonicalize(privateBody) as Record<string, any>) : {};

  const prefix = `${namespace}|v1|`;
  const headersJson = canonicalJsonCompact(headersNormalized);
  const bodyJson = canonicalJsonCompact(bodyNormalized);
  const full = prefix + headersJson + "|" + bodyJson;
  return { aadBytes: enc.encode(full), headersNormalized, bodyNormalized };
}

// Unified transport AAD builder: uses header names verbatim (no canonicalization),
// deep-canonicalizes values and body, and sorts headers by case-insensitive name for stability
export function buildAadFromTransport(
  namespace: string,
  headers: Array<{ header: string; value: any }>,
  body: Record<string, any>
): {
  aadBytes: Uint8Array;
  headersNormalized: Array<{ header: string; value: any }>;
  bodyNormalized: Record<string, any>;
} {
  if (!namespace || namespace.toLowerCase() === "x402") throw new NsForbiddenError("NS_FORBIDDEN");
  const headersNormalized = (headers || []).map((h) => ({
    header: String(h?.header ?? ""),
    value: deepCanonicalize(h?.value),
  }));
  headersNormalized.sort((a, b) => a.header.toLowerCase().localeCompare(b.header.toLowerCase()));
  const bodyNormalized = deepCanonicalize(body || {}) as Record<string, any>;
  const prefix = `${namespace}|v1|`;
  const headersJson = JSON.stringify(headersNormalized);
  const bodyJson = JSON.stringify(bodyNormalized);
  const full = prefix + headersJson + "|" + bodyJson;
  return { aadBytes: enc.encode(full), headersNormalized, bodyNormalized };
}

File: packages/python/x402_hpke/aad.py
import json
from typing import Any, Dict, Tuple, Optional, List
from .errors import NsForbidden, NsCollision
from .extensions import is_approved_extension_header, canonicalize_extension_header


def _deep_canonicalize(obj: Any) -> Any:
    if obj is None or isinstance(obj, (str, int, float, bool)):
        return obj
    if isinstance(obj, list):
        return [_deep_canonicalize(x) for x in obj]
    if isinstance(obj, dict):
        return {k: _deep_canonicalize(obj[k]) for k in sorted(obj.keys())}
    return obj


def _canonical_json(obj: Dict[str, Any]) -> str:
    return json.dumps(_deep_canonicalize(obj), separators=(",", ":"))


def _canonicalize_header_case(h: str) -> str:
    s = str(h or "").lower()
    if s == "x-payment":
        return "X-Payment"
    if s == "x-payment-response":
        return "X-Payment-Response"
    raise ValueError("X402_HEADER")

def validate_x402_core(x: Dict[str, Any]) -> Dict[str, Any]:
    header = _canonicalize_header_case((x or {}).get("header"))
    payload = (x or {}).get("payload")
    if not isinstance(payload, dict) or len(payload.keys()) == 0:
        raise ValueError("X402_PAYLOAD")
    out = {"header": header, "payload": payload}
    for k, v in (x or {}).items():
        if k in ("header", "payload"):
            continue
        out[k] = v
    return out


def build_canonical_aad(
    namespace: str,
    payload: Dict[str, Any],
    extensions: Optional[List[Dict[str, Any]]] = None,
) -> Tuple[bytes, Optional[Dict[str, Any]], Optional[Dict[str, Any]], Optional[Dict[str, Any]], Optional[List[Dict[str, Any]]]]:
    if not namespace or namespace.lower() == "x402":
        raise NsForbidden("NS_FORBIDDEN")
    
    request = payload.get("request")
    response = payload.get("response")
    x402 = payload.get("x402")
    
    primary_json = ""
    x402_normalized, request_normalized, response_normalized = None, None, None

    if x402 is not None:
        x = validate_x402_core(x402)
        primary_json = _canonical_json(x)
        x402_normalized = json.loads(primary_json)
    elif request is not None:
        primary_json = _canonical_json(request)
        request_normalized = json.loads(primary_json)
    elif response is not None:
        primary_json = _canonical_json(response)
        response_normalized = json.loads(primary_json)

    extensions_normalized = None
    extensions_json = ""
    if extensions:
        seen = set()
        norm_exts: List[Dict[str, Any]] = []
        for e in extensions:
            hdr = str((e or {}).get("header") or "")
            if not is_approved_extension_header(hdr):
                raise ValueError("X402_EXTENSION_UNAPPROVED")
            canon_hdr = canonicalize_extension_header(hdr)
            if canon_hdr.lower() in seen:
                raise ValueError("X402_EXTENSION_DUPLICATE")
            ext_payload = (e or {}).get("payload")
            if not isinstance(ext_payload, dict) or len(ext_payload.keys()) == 0:
                raise ValueError("X402_EXTENSION_PAYLOAD")
            ext_obj = {"header": canon_hdr, "payload": ext_payload}
            for k, v in (e or {}).items():
                if k in ("header", "payload"):
                    continue
                ext_obj[k] = v
            norm_exts.append(_deep_canonicalize(ext_obj))
            seen.add(canon_hdr.lower())
        norm_exts.sort(key=lambda x: x.get("header", "").lower())
        extensions_normalized = norm_exts
        extensions_json = _canonical_json(extensions_normalized)
        
    prefix = f"{namespace}|v1|"
    suffix = f"|{extensions_json}" if extensions else "|"
    full = prefix + primary_json + suffix
    return (
        full.encode("utf-8"),
        x402_normalized,
        request_normalized,
        response_normalized,
        extensions_normalized,
    )


def canonical_aad(
    namespace: str,
    payload: Dict[str, Any],
    extensions: Optional[List[Dict[str, Any]]] = None,
) -> bytes:
    return build_canonical_aad(namespace, payload, extensions)[0] 

# New canonical model: headers + body
def canonicalize_core_header_name(h: str) -> str:
    s = str(h or "")
    if s == "":
        return ""
    sl = s.lower()
    if sl == "x-payment":
        return "X-Payment"
    if sl == "x-payment-response":
        return "X-Payment-Response"
    return s

def build_canonical_aad_headers_body(
    namespace: str,
    private_headers: Optional[List[Dict[str, Any]]] = None,
    private_body: Optional[Dict[str, Any]] = None,
) -> Tuple[bytes, List[Dict[str, Any]], Dict[str, Any]]:
    if not namespace or namespace.lower() == "x402":
        raise NsForbidden("NS_FORBIDDEN")
    headers_in = list(private_headers or [])
    seen = set()
    headers_norm: List[Dict[str, Any]] = []
    for e in headers_in:
        raw = str((e or {}).get("header", ""))
        hdr = canonicalize_core_header_name(raw)
        if hdr not in ("X-Payment", "X-Payment-Response", ""):
            if not is_approved_extension_header(hdr):
                raise ValueError("X402_EXTENSION_UNAPPROVED")
            hdr = canonicalize_extension_header(hdr)
        key = hdr.lower()
        if key in seen:
            raise ValueError("X402_EXTENSION_DUPLICATE")
        seen.add(key)
        value = _deep_canonicalize((e or {}).get("value"))
        out = {**(e or {}), "header": hdr, "value": value}
        headers_norm.append(out)
    headers_norm.sort(key=lambda x: str(x.get("header", "")).lower())
    body_norm = _deep_canonicalize(private_body or {})
    prefix = f"{namespace}|v1|"
    headers_json = _canonical_json(headers_norm)
    body_json = _canonical_json(body_norm)
    full = prefix + headers_json + "|" + body_json
    return full.encode("utf-8"), headers_norm, body_norm